### PR TITLE
openssl: bump to 3.0.8

### PR DESCRIPTION
**Pull request by cotequeiroz**
This is a major update to the current LTS version, supported until 2026-09-07.

Changelog:
https://github.com/openssl/openssl/blob/openssl-3.0.8/CHANGES.md

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

---

Here is the list of direct dependent packages:

| Package | Compile-tested | Run-tested | Issues                         |
|-----------| ------------------|-------------|-------------------------|
| libevent2 | OK | | |
| openssl (util) | OK | OK | |
| ustream-ssl| OK | OK | |
| hostapd| OK | OK | |
| adb| OK | | |
| uencrypt| OK | OK | |
| **packages feed** |
| ipmitool | OK | | |
| monit | OK | | |
| rsyslog | OK | | |
| syslog-ng | OK | | |
| zabbix | OK | | |
| cqueues | OK | | |
| erlang | OK | | |
| lua-eco | OK | | |
| lua-openssl | OK | | |
| luaossl | OK | | |
| luasec | OK | | |
| node | OK | | |
| perl-mail-spamassassin | OK | | |
| php8 | OK | | |
| borgbackup | OK | | |
| python-aiohttp | OK | | |
| python-cryptography | OK | | |
| python-curl | OK | | |
| python-docker | OK | | |
| python-dockerpty | OK | | |
| python-engineio | OK | | |
| python-eventlet | OK | | |
| python-paramiko | OK | | |
| python-pyopenssl | OK | | |
| python-pysocks | OK | | |
| python-tornado | OK | | |
| python-twisted | OK | | |
| python-websocket-client | OK | | |
| python3-speedtest-cli | OK | | |
| python3 | OK | | |
| ruby | OK | | |
| afalg_engine | OK | OK* | shows warning when fallback is enabled: expected  |
| apr-util | OK | | |
| cyrus-sasl | OK | | |
| dtndht | OK | | |
| elektra | OK | | |
| getdns | OK | | |
| gost_engine | Fixed | OK | Needs upgrade + cross-compiling work + disabling provider |
| h2o | OK | | |
| ibrcommon | OK | | |
| jose | OK | | |
| ldns | OK | | |
| libarchive | OK | | |
| libesmtp | OK | | |
| libfido2 | OK | | |
| libimobiledevice | OK | | |
| libmariadb | OK | | |
| libnetconf2 | OK | | |
| libnopoll | OK | | |
| libp11 | OK | | |
| libshout | OK | | |
| libssh | OK | | |
| libssh2 | OK | | |
| libstrophe | OK | | |
| libtins | OK | | |
| libtorrent-rasterbar | OK | | |
| libtorrent | OK | | |
| libuhttpd | Fixed | | needs `-Wno-error=deprecated-declarations` |
| libuwsc | OK | | |
| libwebsockets | OK | | |
| libzip | OK | | |
| loudmouth | OK | | |
| neon | OK | | |
| openldap | OK | | |
| opusfile | OK | | |
| poco | OK | | |
| spice | OK | | |
| uw-imap | OK | | |
| alpine | OK | | |
| dovecot | OK | | |
| emailrelay | OK | | |
| exim | OK | | |
| fdm | OK | | |
| mailsend | OK | | |
| mutt | OK | | |
| nail | OK | | |
| opendkim | OK | | |
| postfix | OK | | |
| sendmail | OK | | |
| gst1-plugins-bad | OK | | |
| icecast | OK | | |
| minisatip | OK | | |
| tvheadend | OK | | |
| xupnpd | OK | | |
| aircrack-ng | OK | | |
| apache | OK | | |
| aria2 | OK | | |
| atlas-probe | OK | | |
| bind | OK | | |
| bitlbee | OK | | |
| boinc | Fixed | | invalid conversion from `const rsa_st*` to `rsa_st*`: `-fpermissive` works |
| chaosvpn | OK | | |
| clamav | OK | | |
| conserver | OK | | |
| coova-chilli | OK | | |
| curl | OK | OK | |
| dnsdist | OK | | |
| freeradius3 | OK | | |
| frr | OK | | |
| gatling | OK | | |
| gensio | Unrelated failure | | Compile failure unrelated to openssl: add pcre support to host/swig + add libgensiocpp dependency to python3-gensio |
| git | OK | | |
| gsocket | OK | | |
| haproxy | OK | | |
| hcxdumptool | OK | | |
| hcxtools | OK | | |
| hs20 | OK | | |
| httping | OK | | |
| i2pd | OK | | |
| inadyn | OK | | |
| iperf3 | OK | OK | |
| irssi | OK | | |
| kea | OK | | |
| keepalived | OK | | |
| lftp | OK | | |
| lighttpd | OK | | |
| lynx | OK | | |
| mosh | OK | | |
| mosquitto | OK | | |
| netopeer2 | OK | | |
| netstinky | OK | | |
| nginx-util | Fixed | OK | needs `-Wno-error=deprecated-declarations` |
| nginx | OK | OK | |
| ngircd | OK | | |
| nmap | OK | | |
| nsd | Fixed | | needs a version bump and a fix for building without SSL (fix sent upstream) |
| ntpd | OK | | |
| nut | OK | | |
| open-iscsi | OK | | |
| open-isns | OK | | |
| openconnect | OK | | |
| openfortivpn | OK | | |
| openssh | OK | Client OK | |
| openvpn | OK | | |
| openvswitch | OK | | |
| pagekitec | OK | | |
| pdns-recursor | OK | | |
| pdns | OK | | |
| pen | OK | | |
| prosody | OK | | |
| radsecproxy | OK | | |
| ratched | OK | | |
| rtorrent | OK | | |
| samba4 | OK | | |
| smartdns | OK | | |
| snort | OK | | |
| snort3 | OK | | |
| socat | OK | | |
| softethervpn | OK | | |
| softethervpn5 | OK | | |
| speedtestpp | OK | OK | |
| spoofer | OK | | |
| squid | Fixed | OK | needs major upgrade to 5.7->Wow, it appears to work! |
| sstp-client | OK | | |
| strongswan | OK | | |
| stunnel | OK | | |
| tinc | OK | | |
| tor | OK | | |
| uanytun | OK | | |
| umurmur | Fixed | | `'CRYPTO_MEM_CHECK_ON' undeclared`: this call is used for debugging and can be safely removed |
| unbound | OK | | |
| vallumd | OK | | |
| vpnc | OK | | |
| vsftpd | OK | | |
| wget | OK | OK | |
| znc | OK | | |
| pulseaudio | OK | | |
| shairport-sync | OK | | |
| squeezelite | OK | | |
| apk | OK | | |
| domoticz | OK | | |
| hfsprogs | OK | | |
| lxc | OK | | |
| mariadb | OK | | |
| opensc | OK | | |
| rng-tools | OK | | |
| rtl_433 | OK | | |
| rtty | OK | | |
| ttyd | OK | | |
| usbmuxd | OK | | |
| yara | OK | | |
| **not a direct dependecy, but failing initially** |
| apfree-wifidog | Fixed | | CMake is failing to deal with OpenSSL 3; apply upstream patch |
| **routing feed** |
| hnetd| OK | | |
| **telephony feed** |
| asterisk | OK | | |
| coturn | Fixed | | `'FIPS_mode' undeclared`: needs version bump + patch from gentoo  |
| freeswitch | Fixed | | needs `-Wno-error=deprecated-declarations` |
| kamailio | OK | | |
| libks | OK | | |
| libsrtp | OK | | |
| pjproject | OK | | |
| re | OK | | |
| rtpengine | OK | | |
| signalwire-client-c | OK | | |
| sngrep | OK | | |
| sofia-sip | OK | | |
| yate | OK | | |

### CODE REVIEW

**Comment by dfateyev** [comment]
OpenSSL 3.0.8 is now available, including bug and security fixes.
OpenSSL 3.0 users should upgrade to OpenSSL 3.0.8.
https://www.openssl.org/news/secadv/20230207.txt

**Comment by cotequeiroz** [comment]
I'll update the PR tomorrow.  Thanks.

**Comment by cotequeiroz** [comment]
Ping @aparcar

**Comment by mrkiko** [comment]
Thanks for your work!

**Comment by cotequeiroz** [comment]
Where?

**Comment by Ansuel** [comment]
@cotequeiroz check ci for malta/be

**Comment by cotequeiroz** [comment]
OK.  Most 32-bit arches (but not i386) need it.  Let's see if it works now.

**Comment by cotequeiroz** [comment]
@Ansuel  or anyone familiar with the CI SDK/toolchain being used

CI is failing again for malta/be.  I've been testing with aarch64 only up until now.  It may take me a while to test it locally.

See if my impression is correct, please:
It appears that `libatomic.so.1`  is not part of the external toolchain, the library is copied using its full name `libatomic.so.1.2.0`, not the `dlname` recorded in the dependent library, `libatomic.so.1`, that the linker will look for.

```
$ tar tJvf openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64.tar.xz | grep libatomic
-rwxr-xr-x buildbot/buildbot  1050 2023-02-06 17:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.la
-rwxr-xr-x buildbot/buildbot   79056 2023-02-06 17:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.so.1.2.0
```

**Comment by Ansuel** [comment]
@cotequeiroz interesting it seems the link are not getting migrated on creating the external toolchain tar?

**Comment by cotequeiroz** [comment]
Exactly.

**Comment by Ansuel** [comment]
CI aside from prebuilt tools doesn't do anything special... just use the script to setup external toolchain so this may be a problem in tar creation... 

Should not be a problem to repro

**Comment by Ansuel** [comment]
@cotequeiroz but wait.... the tar have the files so the symbolic link are not getting extracted???

**Comment by cotequeiroz** [comment]
The tar has two libatomic files:  the libtool  text file `libatomic.la`, and the library binary named `libatomic.so.1.2.0`.  It does not have the symlinks to it: `libatomic.so -> libatomic.so.1.2.0`, and more important: `libatomic.so.1 -> libatomic.so.1.2.0`.

**Comment by Ansuel** [comment]
@cotequeiroz from me the output is different

```
tar tJvf openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64.tar.xz | grep libatomic
-rwxr-xr-x buildbot/buildbot  1050 2023-02-06 21:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.la
-rwxr-xr-x buildbot/buildbot   79056 2023-02-06 21:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.so.1.2.0
lrwxrwxrwx buildbot/buildbot       0 2023-02-06 21:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.so -> libatomic.so.1.2.0
lrwxrwxrwx buildbot/buildbot       0 2023-02-06 21:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.so.1 -> libatomic.so.1.2.0
-rw-r--r-- buildbot/buildbot   89942 2023-02-06 21:36 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/lib/libatomic.a```

**Comment by cotequeiroz** [comment]
I compiled all packages, after applying the fixes I mentioned above, and ran some tests with an image on a Xiaomi Redmi Router AX6S (mediatek mt7622/aarch64-cortex-a53) that I have at a non-critical location at home.  I found no major breakage so far.  I'll leave the router running, and will point out anything unusual here.

**Comment by Ansuel** [comment]
Did you by chance have time to try using an external toolchain?

**Comment by cotequeiroz** [comment]
I'm starting this now, using mvebu as target.  I'm trying to build locally here first--more familiar territory--to see if it breaks as well.  It will take a while to get to hostapd still.

**Comment by Ansuel** [comment]
@cotequeiroz sure take your time and thanks for working on this

File: package/libs/openssl/patches/140-allow-prefer-chacha20.patch [hunk 1]
```diff
@@ -76,3 +55,38 @@ Signed-off-by: Eneas U de Queiroz <cote2004-github@yahoo.com>
...
+ {
++#ifdef OPENSSL_PREFER_CHACHA_OVER_GCM
+     return "TLS_AES_256_GCM_SHA384:"
+            "TLS_CHACHA20_POLY1305_SHA256:"
+            "TLS_AES_128_GCM_SHA256";
++#else
++    return "TLS_CHACHA20_POLY1305_SHA256:"
++           "TLS_AES_256_GCM_SHA384:"
++           "TLS_AES_128_GCM_SHA256";
++#endif
```
[the remark this answers is not in the document]
Comment (cotequeiroz) [reply package/libs/openssl/patches/140-allow-prefer-chacha20.patch:70]: You're right, it should.

**Comment by cotequeiroz** [comment]
I think I found why the libatomic is not found:
```
$ readelf -d /home/equeiroz/src/external-toolchain/openwrt/staging_dir/target-mips-openwrt-linux-musl_musl/usr/lib/libcrypto.so

Dynamic section at offset 0x130 contains 36 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libatomic.so.1]
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
 0x0000000e (SONAME)                     Library soname: [libcrypto.so.3]
 0x00000010 (SYMBOLIC)                   0x0
 0x0000000f (RPATH)                      Library rpath: [/home/equeiroz/src/external-toolchain/openwrt/openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin//../../lib:/home/equeiroz/src/external-toolchain/openwrt/openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin//../../usr/lib]
```

There's an extra `../` in the paths.  I need to know where it came from now.

**Comment by cotequeiroz** [comment]
I **think** this was caused by 62c174067616a911ef77a8cf02a6cc2220e534ec.  Then 3e747a7f7534d1c7a5ac1b8e58b99a2585cac0c9 did not fix the everything it intended to.  Can this bug have survived 12.5 years without being detected?

Anyway, the rpath is set by [target/toolchain/files/wrapper.sh](https://github.com/openwrt/openwrt/target/toolchain/files/wrapper.sh), by appending ../.. to the dirname where wrapper.sh was run (via a syslink created by the Makefile)

`wrapper.sh` used to be installed in `usr/bin`, than it was changed to `bin`.  A fix should be coming up shortly.

**Comment by Ansuel** [comment]
@cotequeiroz would be good to understand how this wasn't notice in all this time??? no package ever used this kind of way to find libs?

Anyway guess it lasted 12 years only because external toolchain got fixed only recently and was broken (and unused) for years?

**Comment by cotequeiroz** [comment]
I just took a quick look at it now, and don't have deep knowledge of all this, but I think the reasoning for the ../ mess was to make it relocatable.

**Comment by Ansuel** [comment]
@cotequeiroz anyway if CI is ok I would cherry pick the commit and merge it. Would be ideal if you can create another pr if it's not too annoying for you.

**Comment by cotequeiroz** [comment]
We won't be able to have CI test it before we ship it: wrapper.sh is saved into the toolchain, so CI won't use it until it downloads an SDK that includes the fix.  

```
tar tJvf openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64.tar.xz | grep wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-cpp -> mips-openwrt-linux-musl-wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-as -> mips-openwrt-linux-musl-wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-g++ -> mips-openwrt-linux-musl-wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-gcc -> mips-openwrt-linux-musl-wrapper.sh
-rwxr-xr-x buildbot/buildbot     2440 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-c++ -> mips-openwrt-linux-musl-wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-ld -> mips-openwrt-linux-musl-wrapper.sh
lrwxrwxrwx buildbot/buildbot        0 2023-02-08 23:16 openwrt-toolchain-malta-be_gcc-12.2.0_musl.Linux-x86_64/toolchain-mips_24kc_gcc-12.2.0_musl/bin/mips-openwrt-linux-musl-cc -> mips-openwrt-linux-musl-wrapper.sh
```

Since I could reproduce the exact issue here, I'll generate a toolchain and then try the steps.  I'll definitely open a new PR for it, if it works as expected.

**Comment by cotequeiroz** [comment]
Another change I'll propose is to change `rpath=` to `rpath-link=`.  `rpath` is used to find libraries at run-time, which is not ideal.  I hope nobody ever used an external toolchain under `/tmp` to ship files... ☠️

**Comment by cotequeiroz** [comment]
I'm going all in to get CI to pass...

**Comment by cotequeiroz** [comment]
It worked.  I'll clean this up, open a new PR for the toolchain commits, then I need to take care of the feeds' packages before marking this as ready to merge.

File: package/libs/openssl/Makefile [hunk 2]
```diff
@@ -95,9 +94,10 @@ $(call Package/openssl/Default)
   DEPENDS:=+OPENSSL_WITH_COMPRESSION:zlib \
 	   +OPENSSL_ENGINE_BUILTIN_AFALG:kmod-crypto-user \
 	   +OPENSSL_ENGINE_BUILTIN_DEVCRYPTO:kmod-cryptodev \
-	   +OPENSSL_ENGINE_BUILTIN_PADLOCK:kmod-crypto-hw-padlock
+	   +OPENSSL_ENGINE_BUILTIN_PADLOCK:kmod-crypto-hw-padlock \
+	   +(arm||armeb||mips||mipsel||ppc):libatomic
```
Comment (dhewg) [inline package/libs/openssl/Makefile:98]: IIRC packaged elf files are checked for NEEDED (as per readelf) for missing dependencies.
Does the reverse work? I.e. always add libatomic here, use LDFLAGS+= -Wl,--as-needed and ipkg only adds the dependency if there's a NEEDED entry?

Comment (cotequeiroz) [reply package/libs/openssl/Makefile:98]: Do you mean, just add an unconditional `+libatomic` dependency?
I don't think ipk tries to sort which dependencies are actually needed based on NEEDED entries.
I tried it, and the ipk will have libatomic1 added to Depends: in package for aarch64, even though OpenSSL does not link to libatomic for 64-bit CPUs.
OpenSSL build configuration files decide which targets use libatomic, so `--as-needed` will not make a difference here.

Comment (dhewg) [reply package/libs/openssl/Makefile:98]: Aww, yeah, I meant that. Bummer, that would've been nicer to maintain.
Thanks for trying it out

**Comment by cotequeiroz** [comment]
All done with the feeds, this is ready to go.

**Comment by Ansuel** [comment]
@cotequeiroz the table needs to be updated or we still have some package failing?

**Comment by cotequeiroz** [comment]
They should all be ready for the OpenSSL upgrade.

**Comment by cotequeiroz** [comment]
> Is it difficult to update the package failing for deprecated stuff?

This would depend on the exact API being used.  For example, all of the engine API is now deprecated, in favor of the _provider_ API.  AFAIK OpenSSL has not updated even the engines that ship with it.  This appears to be a difficult task.
If it is just RSA, or other encryption not using the EVP interface (this is the case for some of the packages, IIRC), then it is rather trivial.

I did not even look into it, since we have 5 engine packages (3 in base, 2 in packages feed), which will not be updated in the near future.  We can always reevaluate the situation when such API is dropped in OpenSSL master branch.

**Pull request merged**